modules/src/core/ics04_channel/handler.rs

```diff
@@ -206,9 +208,42 @@ where
         PacketMsg::RecvPacket(msg) => {
             let result = cb.on_recv_packet(module_output, &msg.packet, &msg.signer);
             match result {
-                OnRecvPacketAck::Nil(write_fn) | OnRecvPacketAck::Successful(_, write_fn) => {
+                OnRecvPacketAck::Nil(write_fn) => {
                     write_fn(cb.as_any_mut()).map_err(Error::app_module)?;
                 }
+                OnRecvPacketAck::Successful(acknowledgement, write_fn) => {
+                    write_fn(cb.as_any_mut()).map_err(Error::app_module)?;
+
+                    let type_id = acknowledgement.as_any().type_id();
+                    // process write acknowledgement
+                    if TypeId::of::<TransferAcknowledgement>() == type_id {
```

Comment:
> I don't think this check is required here. Acknowledgement from all app modules must be written.
I think what I understand should be the same, so the App module behind this Acknowledgement used as feedback for processing, if so, this structure needs to be redone with some hierarchical transformation to bring the Acknowledgement to the secondary of the application. I'm not sure it's necessarily the latter that uses Acknowledge, because I've found that the descriptions of such structures in different app specs are not the same structure names
- ICS20 spec `Acknowledgement`: https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#data-structures
- ICS721 spec `Acknowledgement`: https://github.com/cosmos/ibc/tree/master/spec/app/ics-721-nft-transfer#data-structures

Reply:
My understanding is that the core module must treat the acknowledgement as opaque bytes - that is what we call the generic Acknowledgement in ibc-rs (i.e. `core::ics04_channel::msgs::acknowledgement::Acknowledgement`). I just pushed some changes where I modified the `write_acknowledgement::process()` function to now accept that type instead of `Vec<u8>`. We can later improve it to add validation checks (to make sure it is non-empty). This allows us to avoid the `downcast_any()`. Please let me know if you think this could be done better differently.

Reply:
For the time being, I can't think of a better way